src/sliding_sync.rs

```diff
@@ -937,26 +942,79 @@ async fn async_main_loop(
-            if let Some(login_request) = login_receiver.recv().await {
-                log!("Received login request for user {}", login_request.user_id);
-                match login(Cli::from(login_request)).await {
-                    Ok(new_login) => break new_login,
-                    Err(e) => {
-                        error!("Login failed: {e:?}");
-                        Cx::post_action(LoginAction::LoginFailure(e.to_string()));
-                        enqueue_rooms_list_update(RoomsListUpdate::Status {
-                            status: e.to_string(),
-                        });
+        None => {
+            let homeserver_url = url::Url::parse(homeserver_url_str)?;
+            let server_name = ServerName::parse(homeserver_url.domain().unwrap_or("matrix.org"))?;
+            // Obtain the other available login types other than password by first establishing a unauthenticated connection to the server
+            let unauthenticated_client = Client::builder().server_name(&server_name).build().await?;
+            let login_type_res = unauthenticated_client.matrix_auth().get_login_types().await?;
+            let Some(LoginType::Sso(sso_type)) = login_type_res.flows.get(0) else { return Err(anyhow!("Login Type response error")); };
+            Cx::post_action(LoginAction::IdentityProvider(sso_type.identity_providers.clone()));
+            loop {
+                log!("Waiting for login request...");
+                match login_receiver.recv().await {
+                    Some(LoginRequest::LoginByPassword(login_request)) => {
+                        log!("Received password-based login request for user {}", login_request.user_id);
+                        match login(Cli::from(login_request)).await {
+                            Ok(new_login) => break new_login,
+                            Err(e) => {
+                                error!("Login failed: {e:?}");
+                                Cx::post_action(LoginAction::LoginFailure(e.to_string()));
+                                enqueue_rooms_list_update(RoomsListUpdate::Status {
+                                    status: e.to_string(),
+                                });
+                            }
+                        }
+                    }
+                    Some(LoginRequest::LoginBySSO(id)) => {
+                        Cx::post_action(LoginAction::SsoPending(true));
+                        let (client, client_session) =
+                            build_client(&cli, app_data_dir()).await?;
+                        match client
+                            .matrix_auth()
+                            .login_sso(|sso_url: String| async move {
+                                Uri::new(&sso_url).open().map_err(|err| {
+                                    Error::UnknownError(Box::new(io::Error::new(io::ErrorKind::Other, format!(" robius open error {:?}", err))).into())                                        
```

Comment:
"robius open error" is an inappropriate error message because it is merely an implementation detail. Here's a clearer suggestion for a user-facing error.
```suggestion
                                    Error::UnknownError(Box::new(io::Error::new(
                                        io::ErrorKind::Other,
                                        format!("Unable to open SSO login url. Error: {:?}", err)
                                    )).into())                                        
```